gateway/gateway_test.go

```diff
@@ -154,3 +163,72 @@ func testFindServer(data findServerData, t *testing.T) {
 	}
 
 }
+
+type testAction int
+
+const (
+	close testAction = iota
+	receive
+)
+
+func TestBasicGateway(t *testing.T) {
+	hs := handshaking.ServerBoundHandshake{ServerAddress: "infrared"}
+	c1, c2 := net.Pipe()
+	addr := &net.IPAddr{IP: []byte{1, 1, 1, 1}}
+	handshakeConn := connection.NewHandshakeConn(c1, addr)
+	go func() {
+		pk := hs.Marshal()
+		bytes, _ := pk.Marshal()
+		c2.Write(bytes)
+	}()
+	tt := []struct {
+		name   string
+		action testAction
+	}{
+		{
+			name:   "Test or gateway can accept connection",
+			action: receive,
+		},
+		{
+			name:   "Test or gateway can be closed",
+			action: close,
+		},
+	}
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			serverCh := make(chan connection.HandshakeConn)
+			serverData := gateway.ServerData{ConnCh: serverCh}
+			serverStore := &gateway.SingleServerStore{Server: serverData}
+
+			connCh := make(chan connection.HandshakeConn)
+			closeCh := make(chan struct{})
+
+			gw := gateway.NewBasicGatewayWithStore(serverStore, connCh, closeCh)
+			go func() {
+				gw.Start()
+			}()
+
+			switch tc.action {
+			case receive:
+				connCh <- handshakeConn
+				select {
+				case <-time.After(defaultChanTimeout):
+					t.Log("Tasked timed out")
+					t.FailNow()
+				case <-serverCh:
+					t.Log("gateway received and processed connection succesfully")
+				}
+			case close:
+				closeCh <- struct{}{}
+				select {
+				case <-time.After(defaultChanTimeout):
+					t.Log("Everything is fine the task timed out like it should have")
+				case connCh <- handshakeConn:
+					t.Log("Tasked should have timed out")
+					t.FailNow()
+				}
+			}
```

Comment:
There is this problem that if the other action which we wanna do, the result should be the expect reverse. But that will add more complexity to the code itself. So do we wanna have as few lines as possible but more complex to read and understand? or some more lines but easier to understand? Seeing this makes me also realize that previous test also have been focusing more on making stuff compact instead of keeping it easy